src/langchain_google_firestore/document_loader.py

```diff
@@ -0,0 +1,199 @@
+                `page_content`. When only one field is provided only the value is written.
+            metadata_fields: The document field names to write into the `metadata`.
+                By default it will write all fields that are not in `page_content` into `metadata`.
+            client: Client for interacting with the Google Cloud Firestore API.
+        """
+        if client:
+            self.client = client
+            self.client._user_agent = USER_AGENT
+        else:
+            client_info = DEFAULT_CLIENT_INFO
+            client_info.user_agent = USER_AGENT
+            self.client = firestore.Client(client_info=client_info)
+        self.source = source
+        self.page_content_fields = page_content_fields
+        self.metadata_fields = metadata_fields
+
+    def load(self) -> List[Document]:
+        """Load Documents."""
+        return list(self.lazy_load())
+
+    def lazy_load(self) -> Iterator[Document]:
+        """A lazy loader for Documents."""
+        if isinstance(self.source, DocumentReference):
+            self.source._client._client_info.user_agent = USER_AGENT
+            doc = self._load_document()
+            if doc:
+                yield self._load_document()
+            return
+        elif isinstance(self.source, str):
+            query = self.client.collection(self.source)
+        else:
+            query = self.source
+            query._client._client_info.user_agent = USER_AGENT
+
+        for document_snapshot in query.stream():
+            yield DocumentConverter.convertFirestoreDocument(
+                document_snapshot, self.page_content_fields, self.metadata_fields
+            )
+
+    def _load_document(self) -> Document | None:
+        doc = self.source.get()
+        if doc:
+            return DocumentConverter.convertFirestoreDocument(doc)
+        else:
+            return None
+
+
+class FirestoreSaver:
+    """Write into Google Cloud Platform `Firestore`."""
+
+    def __init__(
+        self,
+        collection: Optional[str] = None,
+        client: Optional[Client] = None,
+    ) -> None:
+        """Document Saver for Google Cloud Firestore.
+        Args:
+            collection: The single `/`-delimited path to a Firestore collection. If this
+              value is present it will write documents with an auto generated id.
+            client: Client for interacting with the Google Cloud Firestore API.
+        """
+        self.collection = collection
+        if client:
+            self.client = client
+            self.client._user_agent = USER_AGENT
+        else:
+            client_info = DEFAULT_CLIENT_INFO
+            client_info.user_agent = USER_AGENT
+            self.client = firestore.Client(client_info=client_info)
+
+    def upsert_documents(
+        self,
+        documents: List[Document],
+        merge: Optional[bool] = False,
+        document_ids: Optional[List[str]] = None,
```

Comment:
Why not get document_id from the document.name instead?

Reply:
This is a langchain Document.  The properties are `page_content` and `metadata` neither of those have the name. 

During the conversion I'm setting the path in the `metadata` as     {`reference`: {`path`: `foo/bar`}}

The document_id is an alternative where I will ignore the `metadata` and take the document_id